haystack/components/preprocessors/recursive_splitter.py

```diff
@@ -0,0 +1,408 @@
+                if idx < len(chunks) - 1:
+                    chunks[idx + 1] = remaining_text + (" " if self.split_units == "word" else "") + chunks[idx + 1]
+                elif remaining_text:
+                    # create a new chunk with the trimmed text preceded by the overlap of the last chunk
+                    overlapped_chunks.append(current_chunk)
+                    chunk = remaining_text
+                    overlap, _ = self._get_overlap(overlapped_chunks)
+                    current_chunk = overlap + " " + chunk if self.split_units == "word" else overlap + chunk
+
+            overlapped_chunks.append(current_chunk)
+
+            # it can still be that the new last chunk exceeds the 'split_length'
+            # continue splitting until the last chunk is within 'split_length'
+            if idx == len(chunks) - 1 and self._chunk_length(current_chunk) > self.split_length:
+                last_chunk = overlapped_chunks.pop()
+                first_chunk, remaining_chunk = self._split_chunk(last_chunk)
+                overlapped_chunks.append(first_chunk)
+
+                while remaining_chunk:
+                    # combine overlap with remaining chunk
+                    overlap, _ = self._get_overlap(overlapped_chunks)
+                    current = overlap + (" " if self.split_units == "word" else "") + remaining_chunk
+
+                    # if it fits within split_length we are done
+                    if self._chunk_length(current) <= self.split_length:
+                        overlapped_chunks.append(current)
+                        break
+
+                    # otherwise split it again
+                    first_chunk, remaining_chunk = self._split_chunk(current)
+                    overlapped_chunks.append(first_chunk)
+
+        return overlapped_chunks
+
+    def _get_overlap(self, overlapped_chunks):
+        """Get the previous overlapped chunk instead of the original chunk."""
+        prev_chunk = overlapped_chunks[-1]
+        overlap_start = max(0, self._chunk_length(prev_chunk) - self.split_overlap)
+        if self.split_units == "word":
+            word_chunks = prev_chunk.split()
+            overlap = " ".join(word_chunks[overlap_start:])
+        else:
+            overlap = prev_chunk[overlap_start:]
+        return overlap, prev_chunk
+
+    def _chunk_length(self, text: str) -> int:
+        """
+        Get the length of the chunk in words or characters.
+
+        :param text: The text to be split into chunks.
+        :returns:
+            The length of the chunk in words or characters.
+        """
+        if self.split_units == "word":
+            # page breaks are counted as a single word or page breaks followed by only whitespace 1 or multiple times
+            # regex that matches a page break followed by only whitespace 1 or multiple times
+            if re.match(r"\f\s*", text):
+                return 1
+
+            return len(text.split())
```

Comment:
Instead of using re we can pass the explicit separator to `text.split()` so 
```suggestion
            return len(text.split(" "))
```
passing `" "` causes text.split to properly count the page break. Although it would also count other things like newlines as words whichm might be okay. What do you think?

Reply:
I'm not sure about this, I have to check this again, this changes breaks more tests and I was expecting

Reply:
looking into it, I see your concern with keeping things consistent with the other splitter/chunker

Reply:
Having an issue with this one, the code interprets each separator as a regex that also captures the separator itself. It then adds it to the previous unit token so that it's not lost and we can have it when constructing the chunks. 

This is working for all cases, even with this new way of counting (i.e., `text.split(" ")`.

Now, the issue is when I'm separating by seperator=[" "], this will result in something like this:

```python
text = "This is some text. \f This text is on another page. \f This is the last pag3."

['This ',
 'is ',
 'some ',
 'text. ',
 '\x0c ',
 'This ',
 'text ',
 'is ',
 'on ',
 'another ',
 'page. ',
 '\x0c ',
 'This ',
 'is ',
 'the ',
 'last ',
 'pag3.']
```

This thing is that now the white space is also counted as a character, i.e.:

```python
In [8]: len(splits[0].split(" "))
Out[8]: 2

In [9]: len(splits[0].split())
Out[9]: 1
```

Then, if we specify we want chunks of size=4, it builds something like `This is `. 

I'm working on a solution to handle this specific type of separator, and way of counting, so that it can be counted as `text.split()`, meaning that white spaces are not counted and we can still have them when constructing the chunk based from the original string.

Right now, it involves having an extra parameter in for the `self._chunk_length()` informing which current separator we are using and then avoiding counting the white spaces.

Reply:
Unfortunately, this also doesn't seem to work, also tried with the `regex` library, and the core `re`.

```python
splits = ['This ',
 'is ',
 'some ',
 'text. ',
 '\x0c ',
 'This ',
 'text ',
 'is ',
 'on ',
 'another ',
 'page. ',
 '\x0c ',
 'This ',
 'is ',
 'the ',
 'last ',
 'pag3.']

import regex

len(regex.split(" ", splits[0]))
Out[61]: 2

import re
len(re.split(" ", splits[0]))
Out[64]: 2
```



Another possible solution would be something like this, but the issue is to have an exhaustive list of `special_chars` that covers most use cases.

```python
def count_words(text):
    words = [word for word in text.split() if word]
    special_chars = text.count('\n') + text.count('\f') + text.count('\x0c')
    return len(words) + special_chars
```